File: implement.js
// Time Complexity : O(N) //number of characters
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/**
 * Initialize your data structure here.
 */
var Trie = function() {
    this.children = new Map();
    this.isEnd = false;
};

/**
 * Inserts a word into the trie. 
 * @param {string} word
 * @return {void}
 */
Trie.prototype.insert = function(word) {
    let curr = this
    for(let i = 0; i < word.length; i++){
        let node = curr.children.get(word[i])
        if(!node){
            node = new Trie();
            curr.children.set(word[i], node);
        }
        curr = node
    }
    curr.isEnd = true;
};

/**
 * Returns if the word is in the trie. 
 * @param {string} word
 * @return {boolean}
 */
Trie.prototype.search = function(word) {
    let curr = this;
    for(let i = 0; i < word.length; i++){
        if(!curr.children.has(word[i])) return false
        let node = curr.children.get(word[i])
        if(!node) return false
        curr = node
    }
    return curr.isEnd;
    
};

/**
 * Returns if there is any word in the trie that starts with the given prefix. 
 * @param {string} prefix
 * @return {boolean}
 */
Trie.prototype.startsWith = function(prefix) {
    let curr = this;
    for(let i = 0; i < prefix.length; i++){
        if(!curr.children.has(prefix[i])) return false
        let node = curr.children.get(prefix[i])
        if(!node) return false
        curr = node
    }
    return true;
};

/** 
 * Your Trie object will be instantiated and called as such:
 * var obj = new Trie()
 * obj.insert(word)
 * var param_2 = obj.search(word)
 * var param_3 = obj.startsWith(prefix)
 */

File: Implement.java
//Time complexity-O(max length of string)
//Space complexity-O(length of unique prefix + number of suffix)
//Ran on leetcode:Yes
//Solution with comments-
class Trie {
        class TrieNode{
            TrieNode[] children;
            boolean isValid;
            
            public TrieNode(){
                children= new TrieNode[26];
                isValid=false;
            }
           
        }
    
     TrieNode root;
    /** Initialize your data structure here. */
    public Trie() {
       root = new TrieNode();
    }
    
    /** Inserts a word into the trie. */
    public void insert(String word) {
        TrieNode curr= root;
        for(char c : word.toCharArray()){
            if(curr.children[c-'a']==null){//Index of trie is charcter's numeric value. 
                curr.children[c-'a']=new TrieNode();
            }
            curr= curr.children[c-'a'];
        }
        curr.isValid=true;//valid word only when we are at the end of the word.
    }
    
    /** Returns if the word is in the trie. */
    public boolean search(String word) {
        TrieNode curr= root;
         for(char c : word.toCharArray()){
            if(curr.children[c-'a']==null){
                return false;
            }
            curr=curr.children[c-'a'];
        }
        
        return curr.isValid;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String prefix) {
        TrieNode curr= root;
        for(char c : prefix.toCharArray()){
            if(curr.children[c-'a']!=null){
                curr=curr.children[c-'a'];
            }
            else{
                return false;
                }
        }
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */s